### ISSUE TITLE
ORC + GC_ref/unref + closure + destructor lifetime issue

### ISSUE THREAD

**Issue by yglukhov** [nim-lang/Nim#17506]
```nim
type
  Foo = ref FooObj
  FooPtr = ptr FooObj
  FooObj = object
    cb: proc()

proc `=destroy`(f: var FooObj) =
  echo "destroy!"
  # The sample will work if next line removed. Why?
  # Even though destructor will still be called too early!
  echo "des: ", cast[int](addr f)

proc clear(f: Foo) =
  f.cb = nil
  GC_unref(f)

proc makeFoo(): FooPtr =
  let f = Foo()
  GC_ref(f)
  f.cb = proc() =
    let a = cast[int](f)
    clear(f)
    # Here we're expecting f to be valid, but it's not
    let b = cast[int](f)
    doAssert(b == a) # FAILURE HERE!
  result = cast[FooPtr](f)

proc test() =
  let f = makeFoo()
  f.cb()

test()
```
```
nim c -r --gc:orc test.nim
```
**Expected result:** runs without errors
**Actual result:** assertion fails

**Comment by Araq** [comment]
After you call `clear` the memory is freed immediately, the line `let b = cast[int](f)` does not keep f alive nor should it, IMHO. The workaround is rather clean too:

```nim

type
  Foo = ref FooObj
  FooPtr = ptr FooObj
  FooObj = object
    cb: proc()

proc `=destroy`(f: var FooObj) =
  echo "destroy!"
  # The sample will work if next line removed. Why?
  # Even though destructor will still be called too early!
  echo "des: ", cast[int](addr f)

proc clear(f: Foo) =
  f.cb = nil
  GC_unref(f)

proc makeFoo(): FooPtr =
  let f = Foo()
  GC_ref(f)
  f.cb = proc() =
    let a = cast[int](f)

    # use 'f' before you clear it out:
    let b = cast[int](f)
    doAssert(b == a) # FAILURE HERE!

    clear(f)
  result = cast[FooPtr](f)

proc test() =
  let f = makeFoo()
  f.cb()

test()
```

Feel free to reopen if you disagree.

**Comment by yglukhov** [comment]
EDIT: Ignore this comment

I'm sorry, I missed your response, but I disagree. Let me prove it differently. Consider this slightly simplified example, basically `clear` function is now inlined:
```nim
type
  Foo = ref FooObj
  FooPtr = ptr FooObj
  FooObj = object
    cb: proc()

proc `=destroy`(f: var FooObj) =
  echo "destroy!"
  # The sample will work if next line removed. Why?
  # Even though destructor will still be called too early!
  echo "des: ", cast[int](addr f)

proc makeFoo(): FooPtr =
  let f = Foo()
  # GC_ref(f)
  f.cb = proc() =
    let a = cast[int](f)
    f.cb = nil
    # GC_unref(f)
    # Here we're expecting f to be valid, but it's not
    let b = cast[int](f)
    doAssert(b == a) # FAILURE HERE!
  result = cast[FooPtr](f)

proc test() =
  let f = makeFoo()
  f.cb()

test()
```
- The code above is valid, isn't it? And it works as expected.
- Now if we uncomment `GC_ref` and `GC_unref` that should not change anything, because they're balanced out, right? But it actually breaks the code. In real world scenarios `GC_ref` and `GC_unref` may not be controlled by the author of this code, so composability is broken at least.

Does it make sense?

**Comment by yglukhov** [comment]
> The code above is valid, isn't it?

My bad, it's not valid as FooPtr outlives Foo, but somehow it still works :)

> the line `let b = cast[int](f)` does not keep f alive nor should it, IMHO

(EDIT: Ignore this) So what is the way to keep `b` alive in scope?

**Comment by yglukhov** [comment]
It seems I've found a solid solution for this. If I understood it correctly, the issue happens because closure environment is destroyed when the closure is being executed. I found, that it happens when the closure is called as `f.cb()` where `f` is `ptr object`. Making `f` a `ref object` will fix the issue, as nim will emit additional refcounts to keep it alive.
```nim
let p: ptr Foo = ...
p.cb() # Unsafe! Might crash!
let r: ref Foo = cast[ref Foo](p)
r.cb() # Good!
```
So all of this issue boils down to "it may be unsafe (UB?) to call a member closure of a ptr object", which I'm fine with.
